core/go/internal/sequencer/coordinator/coordinator.go

```diff
@@ -0,0 +1,620 @@
+				c.inFlightMutex.Wait()
+			}
+
+			// Dispatch and then asynchronously update the state machine to State_Dispatched
+			log.L(ctx).Debugf("submitting transaction %s for dispatch", tx.ID.String())
+			c.readyForDispatch(ctx, tx)
+
+			// Dispatched transactions that result in a chained private transaction don't count towards max dispatch ahead
+			if tx.PreparedPrivateTransaction == nil {
+				dispatchedAhead++
+			}
+
+			// Update the TX state machine
+			c.QueueEvent(ctx, &transaction.DispatchedEvent{
+				BaseCoordinatorEvent: transaction.BaseCoordinatorEvent{
+					TransactionID: tx.ID,
+				},
+			})
+
+			// We almost never need to wait for the state machine's event loop to process the update to State_Dispatched
+			// but if we hit the max dispatch ahead limit after dispatching this transaction we do, because we can't be sure
+			// in-flight will be accurate on the next loop round
+			if len(c.inFlightTxns)+dispatchedAhead >= c.maxDispatchAhead {
+				for c.inFlightTxns[tx.ID] == nil {
+					c.inFlightMutex.Wait()
+				}
+				dispatchedAhead = 0
+			}
+			c.inFlightMutex.L.Unlock()
+		case <-c.stopDispatchLoop:
+			log.L(ctx).Infof("coordinator dispatch loop stopped")
+			return
+		case <-ctx.Done():
+			log.L(ctx).Infof("coordinator dispatch loop cancelled")
+			return
+		}
+	}
+}
+
+func (c *coordinator) sendHandoverRequest(ctx context.Context) {
+	err := c.transportWriter.SendHandoverRequest(ctx, c.activeCoordinatorNode, c.contractAddress)
+	if err != nil {
+		log.L(ctx).Errorf("error sending handover request: %v", err)
+	}
+}
+
+func (c *coordinator) GetActiveCoordinatorNode(ctx context.Context, initIfNoActiveCoordinator bool) string {
+	if initIfNoActiveCoordinator && c.activeCoordinatorNode == "" {
+		// If we don't yet have an active coordinator, select one based on the appropriate algorithm for the contract type
+		activeCoordinator, err := c.SelectActiveCoordinatorNode(ctx)
+		if err != nil {
+			log.L(ctx).Errorf("error selecting next active coordinator: %v", err)
+			return ""
+		}
+		c.activeCoordinatorNode = activeCoordinator
+	}
+	return c.activeCoordinatorNode
+}
+
+func (c *coordinator) SelectActiveCoordinatorNode(ctx context.Context) (string, error) {
+	coordinator := ""
+	if c.domainAPI.ContractConfig().GetCoordinatorSelection() == prototk.ContractConfig_COORDINATOR_STATIC {
+		// E.g. Noto
+		if c.domainAPI.ContractConfig().GetStaticCoordinator() == "" {
+			return "", fmt.Errorf("static coordinator mode is configured but static coordinator node is not set")
+		}
+		log.L(ctx).Debugf("coordinator %s selected as next active coordinator in static coordinator mode", c.domainAPI.ContractConfig().GetStaticCoordinator())
+		// If the static coordinator returns a fully qualified identity extract just the node name
+		if strings.Contains(c.domainAPI.ContractConfig().GetStaticCoordinator(), "@") {
+			parts := strings.Split(c.domainAPI.ContractConfig().GetStaticCoordinator(), "@")
```

Comment:
This identity parsing code exists in exactly one place, and must be used from there.